File: daily/32 (Longest valid Parenthesis)/Solution.java
class Solution{
    public int LongestValidParentheses(String s) {
        int maxLength = 0;
        int stack[] = new int[s.length()];
        int top = -1;
        int lastInvalid = -1;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                stack[++top] = i; 
            } else {
                if (top == -1) {
                    lastInvalid = i; 
                } else {
                    top--; 
                    if (top == -1) {
                        maxLength = Math.max(maxLength, i - lastInvalid); 
                    } else {
                        maxLength = Math.max(maxLength, i - stack[top]);
                    }
                }
            }
        }

        
        return maxLength;
    }
    public static void main(String[] args) {
        Solution solution = new Solution();
        String s = ")(";
        int result = solution.LongestValidParentheses(s);
        System.out.println("Longest valid parentheses length: " + result); 
    }
}

File: daily/32 (Longest valid Parenthesis)/Solution.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        int maxLength = 0;
        int left = 0, right = 0;
        for (char c : s) {
            if (c == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                maxLength = max(maxLength, 2 * right);
            } else if (right > left) {
                left = right = 0; 
            }
        }
        left = right = 0;
        for (int i = s.size() - 1; i >= 0; i--) {
            if (s[i] == ')') {
                right++;
            } else {
                left++;
            }
            if (left == right) {
                maxLength = max(maxLength, 2 * left);
            } else if (left > right) {
                left = right = 0; 
            }
        }

        return maxLength;
    }
};

int main() {
    Solution solution;
    string s = "((";
    cout << "Longest valid parentheses length: " << solution.longestValidParentheses(s) << endl;
    return 0;
}